app/__init__.py

```diff
@@ -65,6 +59,91 @@
     "How many concurrent requests are currently being served",
 )
 
+local_resetter = LazyLocalGetterResetter()
+
+#
+# "clients" that need thread-local copies
+#
+
+_firetext_client_context_var: ContextVar[FiretextClient] = ContextVar("firetext_client")
+get_firetext_client: LazyLocalGetter[FiretextClient] = LazyLocalGetter(
+    _firetext_client_context_var,
+    lambda: FiretextClient(current_app, statsd_client=statsd_client),
+    expected_type=FiretextClient,
+)
+local_resetter.register(get_firetext_client)
+firetext_client = LocalProxy(get_firetext_client)
+
+_mmg_client_context_var: ContextVar[MMGClient] = ContextVar("mmg_client")
+get_mmg_client: LazyLocalGetter[MMGClient] = LazyLocalGetter(
+    _mmg_client_context_var,
+    lambda: MMGClient(current_app, statsd_client=statsd_client),
+    expected_type=MMGClient,
+)
+local_resetter.register(get_mmg_client)
+mmg_client = LocalProxy(get_mmg_client)
+
+_aws_ses_client_context_var: ContextVar[AwsSesClient] = ContextVar("aws_ses_client")
+get_aws_ses_client: LazyLocalGetter[AwsSesClient] = LazyLocalGetter(
+    _aws_ses_client_context_var,
+    lambda: AwsSesClient(current_app.config["AWS_REGION"], statsd_client=statsd_client),
+    expected_type=AwsSesClient,
+)
+local_resetter.register(get_aws_ses_client)
+aws_ses_client = LocalProxy(get_aws_ses_client)
+
+_aws_ses_stub_client_context_var: ContextVar[AwsSesStubClient] = ContextVar("aws_ses_stub_client")
+get_aws_ses_stub_client: LazyLocalGetter[AwsSesStubClient] = LazyLocalGetter(
+    _aws_ses_stub_client_context_var,
+    lambda: AwsSesStubClient(
+        current_app.config["AWS_REGION"],
+        statsd_client=statsd_client,
+        stub_url=current_app.config["SES_STUB_URL"],
+    ),
+    expected_type=AwsSesStubClient,
+)
+local_resetter.register(get_aws_ses_stub_client)
+aws_ses_stub_client = LocalProxy(get_aws_ses_stub_client)
+
+_notification_provider_clients_context_var: ContextVar[NotificationProviderClients] = ContextVar(
+    "notification_provider_clients"
+)
+get_notification_provider_clients: LazyLocalGetter[NotificationProviderClients] = LazyLocalGetter(
+    _notification_provider_clients_context_var,
+    lambda: NotificationProviderClients(
+        sms_clients={
+            getter.expected_type.name: LocalProxy(getter)
+            for getter in (
+                get_firetext_client,
+                get_mmg_client,
+            )
+        },
+        email_clients={
+            getter.expected_type.name: LocalProxy(getter)
+            for getter in ((get_aws_ses_stub_client,) if current_app.config["SES_STUB_URL"] else (get_aws_ses_client,))
+        },
+    ),
```

Comment:
I think I need an explanation of what's happening here, I don't really understand

Reply:
Two dictionary comprehensions, generating dicts of the `expected_type`'s `.name` mapped against a `LocalProxy` to its getter.

The `email_clients` one switches its getter based on `SES_STUB_URL`.